Objects/genericaliasobject.c

```diff
@@ -320,20 +320,85 @@ subs_tvars(PyObject *obj, PyObject *params,
 static int
 _is_unpacked_typevartuple(PyObject *arg)
 {
-    PyObject *meth;
-    int res = _PyObject_LookupAttr(arg, &_Py_ID(__typing_unpacked__), &meth);
+    PyObject *tmp;
+    if (PyType_Check(arg)) { // TODO: Add test
+        return 0;
+    }
+    int res = _PyObject_LookupAttr(arg, &_Py_ID(__typing_is_unpacked_typevartuple__), &tmp);
     if (res > 0) {
-        PyObject *tmp = PyObject_CallNoArgs(meth);
-        Py_DECREF(meth);
-        if (tmp == NULL) {
-            return -1;
-        }
         res = PyObject_IsTrue(tmp);
         Py_DECREF(tmp);
     }
     return res;
 }
 
+static PyObject *
+_unpacked_tuple_args(PyObject *arg)
+{
+    PyObject *result;
+    assert(!PyType_Check(arg));
+    // Fast path
+    if (_PyGenericAlias_Check(arg) &&
+            ((gaobject *)arg)->starred &&
+            ((gaobject *)arg)->origin == (PyObject *)&PyTuple_Type)
+    {
+        result = ((gaobject *)arg)->args;
+        Py_INCREF(result);
+        return result;
+    }
+
+    if (_PyObject_LookupAttr(arg, &_Py_ID(__typing_unpacked_tuple_args__), &result) > 0) {
+        if (result == Py_None) {
+            Py_DECREF(result);
+            return NULL;
+        }
+        return result;
+    }
+    return NULL;
+}
+
+static PyObject *
+_unpack_args(PyObject *item)
+{
+    PyObject *newargs = PyList_New(0);
+    if (newargs == NULL) {
+        return NULL;
+    }
+    int is_tuple = PyTuple_Check(item);
+    Py_ssize_t nitems = is_tuple ? PyTuple_GET_SIZE(item) : 1;
+    PyObject **argitems = is_tuple ? &PyTuple_GET_ITEM(item, 0) : &item;
+    for (Py_ssize_t i = 0; i < nitems; i++) {
+        item = argitems[i];
+        if (!PyType_Check(item)) {
+            PyObject *subargs = _unpacked_tuple_args(item);
+            if (subargs != NULL &&
+                PyTuple_Check(subargs) &&
+                !(PyTuple_GET_SIZE(subargs) &&
+                  PyTuple_GET_ITEM(subargs, PyTuple_GET_SIZE(subargs)-1) == Py_Ellipsis))
+            {
+                if (PyList_SetSlice(newargs, PY_SSIZE_T_MAX, PY_SSIZE_T_MAX, subargs) < 0) {
```

Comment:
Noob question: how does setting it to `PY_SSIZE_T_MAX` work? I'm surprised it's not something like `low=i`, `high=i+PyTuple_GET_SIZE(subargs)` ? I think I missed something.

Reply:
It works like:
```python
a = [1, 2, 3]
a[sys.maxsize:sys.maxsize] = [4, 5]
```

Reply:
I knew the Python equivalent but my mind is blown. I didn't know that was possible in Python! I'd always assumed `list.extend` was the only way to do it. Thanks for teaching me something new today Serhiy.

Last question: Why not use `_PyList_Extend`. Or is it just a personal preference which to use?

Reply:
`_PyList_Extend` is not convenient. It returns `Py_None`, for which you should call `Py_DECREF`. Compare:
```c
if (PyList_SetSlice(newargs, PY_SSIZE_T_MAX, PY_SSIZE_T_MAX, subargs) < 0) {
    // error
}
```
and
```c
PyObject *tmp = _PyList_Extend(newargs, subargs);
if (tmp == NULL) {
    // error
}
else {
    Py_DECREF(tmp);
}
```
Using `PyList_SetSlice()` for `list.extend()` or `list.clear()` is idiomatic.